### DISCUSSION TITLE
Add an alternative StreamingResponse class which can automatically encode to JSON Pydantic entities coming from a generator

### DISCUSSION THREAD

**Discussion by andreagrandi** [fastapi/fastapi#9120] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Implement a Pull Request for this feature if maintainers are interested in

### Example

Here's a self-contained [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi.responses import JSONEncodedStreamingResponse
from pydantic import BaseModel
from datetime import datetime
from fastapi.encoders import jsonable_encoder
import json

app = FastAPI()

class MyEntity(BaseModel):
    id: int
    name: str
    joined_at: datetime

def generate_results():
    for i in range(10):
        yield MyEntity(
            id=i,
            name=f"name_{i}",
            joined=datetime.now(),
        )

@app.get("/")
async def stream_data():
    results = generate_results()
    return JSONEncodedStreamingResponse(results)
```

instead of this thing I have to do now:

```Python
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from datetime import datetime
from fastapi.encoders import jsonable_encoder
import json

app = FastAPI()

class MyEntity(BaseModel):
    id: int
    name: str
    joined_at: datetime

def generate_results():
    for i in range(10):
        yield MyEntity(
            id=i,
            name=f"name_{i}",
            joined=datetime.now(),
        )

@app.get("/")
async def stream_data():
    results = generate_results()

    def _encoded_results():
        yield "["
        for idx, item in enumerate(results):
            if idx > 0:
                yield ","
            yield json.dumps(jsonable_encoder(item.dict()))
        yield "]"

    return StreamingResponse(_encoded_results())
```

**Note**: I haven't double checked if the above examples run "as is", but it's not important, it's just a stub to give you an idea of what I would like to do.

### Description

I've a method that is going to return probably more than 10.000 results from a SQLAlchemy query (but the ORM I use is not relevant here), which are transformed to Pydantic entities. These entities are yield-ed to the web handler, but unfortunately the `StreamingResponse` can't stream them "as is" (hint: https://github.com/encode/starlette/issues/419#issuecomment-470077657 ).

The workaround I've found essentially must transform the data from Pydantic entities using `.dict()` first, passing through a `jsonable_encoder` and finally calling `json.dumps(...)` for something like this:

```python
def _encoded_results():
        yield "["
        for idx, item in enumerate(results):
            if idx > 0:
                yield ","
            yield json.dumps(jsonable_encoder(item.dict()))
        yield "]"
```

### The solution you would like

Since the above `_encoded_results()` would be the same for any Pydantic entity, I was wondering why not having a `JSONEncodedStreamingResponse` (I'm up for naming this in a different way if you can find a better name!) that internally does it for me? So I don't have to repeat that code around.

### Describe alternatives you've considered

Keep using the workaround described above, but it's really awful.

### Environment

* OS: any
* FastAPI Version: 0.52.0
* Python version: Python 3.8.1

### Additional context

I'm happy to work on this feature/PR, since I would like to contribute back to this library, but before spending any additional time to make it work nicely I'd like to understand if there is interest on your side and if you think this is something that should live inside FastAPI or as an external library.

**Comment by NoahCardoza** [comment]
Supporting `ndjson` might be a solution to this issue (https://github.com/tiangolo/fastapi/issues/2039). Kill two birds with one enhancement.

**Comment by padraic-shafer** [comment]
> ### The solution you would like
> Since the above `_encoded_results()` would be the same for any Pydantic entity, I was wondering why not having a `JSONEncodedStreamingResponse` (I'm up for naming this in a different way if you can find a better name!) that internally does it for me? So I don't have to repeat that code around.

@andreagrandi:  I incorporated the gist of your example above by subclassing `StreamingResponse` in one of my projects. The supplied generator can be synchronous or asynchronous.

I chose to call it `StreamingJsonListResponse` because the iterable content is returned as a list.  For other use cases, I could imagine supplying a generator that returns (key, value) pairs that should instead return a JSON stream with dict-like mapping; and probably there are other common iterator paradigms that would be useful.  If that is true, then perhaps it would be better to define the `_encoded_async_generator()` and `_encoded_generator()` functions in an abstract mix-in class for a more generic `StreamingJsonResponse` class.

```python3
from typing import (
    Iterable, AsyncIterable,
    Optional, List, Mapping, Union, Literal, Any,
)
# from pydantic import BaseModel
from starlette.responses import StreamingResponse
from starlette.background import BackgroundTask
from starlette.concurrency import iterate_in_threadpool
from fastapi.encoders import jsonable_encoder
import json


async def async_enumerate(async_sequence, start=0):
    """Asynchronously enumerate an async iterator from a given start value.

    See https://stackoverflow.com/a/55930068/9639441
    """
    idx = start
    async for element in async_sequence:
        yield idx, element
        idx += 1


class StreamingJsonListResponse(StreamingResponse):
    """Converts a pydantic model generator into a streaming HTTP Response
    that streams a JSON list, one element at a time.
    
    See https://github.com/tiangolo/fastapi/issues/1978
    """
    def __init__(
        self,
        content_generator: Iterable,
        status_code: int = 200,
        headers: Optional[Mapping[str, str]] = None,
        media_type: Optional[str] = None,
        background: Optional[BackgroundTask] = None,
    ) -> None:
        if isinstance(content_generator, AsyncIterable):
            body_iterator = self._encoded_async_generator(content_generator)
        else:
            body_iterator = self._encoded_generator(content_generator)
        super().__init__(
            content=body_iterator,
            status_code=status_code,
            headers=headers,
            media_type=media_type,
            background=background,
        )

    async def _encoded_async_generator(self, async_generator):
        """Converts an asynchronous pydantic model generator 
            into a streaming JSON list
        """
        yield "["
        async for idx, item in async_enumerate(async_generator):
            if idx > 0:
                yield ","
            yield json.dumps(jsonable_encoder(item.dict()))
        yield "]"

    async def _encoded_generator(self, generator):
        """Converts a synchronous pydantic model generator 
            into a streaming JSON list
        """
        yield "["
        for idx, item in enumerate(generator):
            if idx > 0:
                yield ","
            yield json.dumps(jsonable_encoder(item.dict()))
        yield "]"
```
